### DISCUSSION TITLE
Retrofitting ESP8266 devices + the ESP32­-C3-­MINI-­1

### DISCUSSION THREAD

**Discussion by techman83** [arendst/Tasmota#13761] [Show and tell]
I've been ordering random ESP32-C3 modules to assess how well my older IoT devices could be upgraded. I've played around with some carrier boards, but where the slotted style packages (ie the Tuya TYWE boards) fit, there tends to be not much room. These actually came with a bunch of WT32C3-O1N chips I bought to see how easy they'd be to retrofit into some Tuya based downlights.

Datasheet -> https://www.espressif.com/sites/default/files/documentation/esp32-c3-mini-1_datasheet_en.pdf

Front:
![2021-11-22 13 40 08](https://user-images.githubusercontent.com/3467784/142807625-cb167e96-6daa-4399-a694-584eaf50eb00.jpg)

Back:
![2021-11-22 13 39 57](https://user-images.githubusercontent.com/3467784/142807626-fd072b72-3c13-4194-a33b-19233c872e89.jpg)

In-between a TYWE2L and a WT32C3-O1N
![2021-11-22 13 40 36](https://user-images.githubusercontent.com/3467784/142807727-787c918c-147a-4973-9da9-25da581d460a.jpg)

I think it'd be relatively straight forward to build a suite of PCB adapters that could make for drop in retrofits for the ESP8266 based devices that are out there. I have frankensteined a few things, but having an adapter for these would be way easier. I'll share if design something, but relatively time poor at the moment!

Custom carrier board  - it's a pain to solder up and adds too much height (plus I got the pins slightly wrong :stuck_out_tongue:):
![2021-11-22 13 55 10-1](https://user-images.githubusercontent.com/3467784/142808476-76a8898c-5b9a-4f0e-8e7b-30644e460eb2.jpg)

Hacking an ESP32-O1M into a smart plug. Required some careful slotting with a hacksaw blade and a bodge wire:
![2021-10-21 15 53 25](https://user-images.githubusercontent.com/3467784/142808479-fbb5ef67-edcc-4a14-87c7-9955c151656d.jpg)

**Comment by sfromis** [comment]
For retrofitting, I can also recommend the ESP-C3-12F module, which has the same layout as ESP-12 (or the many other modules having that form factor), no adapter needed.
![image](https://user-images.githubusercontent.com/47082390/142830287-d4f40267-1a1f-4871-b036-560905d1b8a9.png)

**Reply by techman83** [reply]
I have a bunch of those, but I don't actually have many devices that use a 12F package.

**Reply by sfromis** [reply]
I've seen many commercial devices use the ESP-12 form factor (not specifically ESP-12F with 6 extra pins, not needed). One example is WB3S.

**Reply by techman83** [reply]
Yeah I do have a couple with the ESP-12, but overwhelmingly most of mine are the TWYE2L form factor and I've been hunting down a retrofit solution to experiment with. One that is more of a drop in, rather than some of the more extreme/time consuming mods I've tried so far.

**Comment by techman83** [comment]
Looks to be a pretty new release -> https://www.cnx-software.com/2021/11/22/meet-esp-c3-m1-and-esp-c3-m1-i-ultra-small-esp32-c3-mini-modules/

**Comment by digiblur** [comment]
There's also these if you dig around https://wireless-tag.en.alibaba.com/product/1600348544006-828900247/WT32C3_01N_4MB_OEM_ESP32_wi_fi_fi_WIFI_BLE_5_0_LE_Module_with_ESP32_C3_Chip.html?spm=a2700.shop_plgr.89.4.34cd5847VMrLi3

**Reply by techman83** [reply]
Yes, I have a bunch of those. One is in the right of the comparison pictured above :smile: In fact, the samples I received came with that order!

**Reply by techman83** [reply]
I've done a heap of retrofitting with these now, so I've added the pin translation to the wiki page

https://templates.blakadder.com/WT32C3-01N.html

I should upload my flashing jig to Thingiverse and update this thread with some retrofitting pictures. With a bit of flux + solder braid, and some practice, the process is pretty fast!

**Reply by dori4n** [reply]
I just got a bunch of these [Wireless Tag WT32C3-01N V1.3](https://templates.blakadder.com/WT32C3-01N.html "") modules in the mail.
Do you know how to program these?
I'd like to replace a [CB2S](https://developer.tuya.com/en/docs/iot/cb2s-module-datasheet "CB2S Module Datasheet - Tuya") in a Wi-Fi Smart Plug Outlet with it after installing Tasmota on it.
I can't find a programming socket for this.
Someone [flashed ESPHome](https://community.home-assistant.io/t/tuya-wifi-smart-socket-cb2s-esphome-configuration/719413) to the Tuya module, but I'd prefer not soldering them in before trying to flash and testing them.
Is there some type of common card edge connector available for this that I can solder onto an adapter or stick into a breadboard that would allow me to slot the modules in like a game cartridge in a game console from the 1990s?

**Reply by sfromis** [reply]
Connections are already on the templates page you linked (including the data sheet linked there). The usual tx/rx for flashing, along with power and enable, are available on the edge connector, but do remember the part about pulling gpio 9 low (similar to gpio 0 for ESP8266). And the last one is not available on the edge connector.

**Reply by dori4n** [reply]
> Connections are already on the templates page you linked (including the data sheet linked there). 

Except, you will note, that I asked about how to program these, not their pinout. This also requires some configuration, and since some of you did this already, I would have expected you know already and can tell me so at least it becomes somewhat publicly documented.

I also asked if you know about a programming *socket* that allows press-fitting the module for programming. Similar to these for the ESP8266 and ESP32: [eBay](https://www.ebay.com/itm/404255088515), [AliExpress](https://www.aliexpress.com/i/1005004888105578.html), [Amazon](https://www.amazon.com/dp/B0BPCD9C83/)

**Reply by sfromis** [reply]
How to program chips is already on the regular [getting-started](https://tasmota.github.io/docs/Getting-Started/) page in the docs. The main difference with ESP32-C3 is that gpio 9 is to be grounded for flashing mode, instead of gpio 0 on most other Espressif chips. 

But that form factor is a bit tricky, which is why I pointed out that you cannot do flashing only using the pins on the edge connector, as gpio 9 is needed too. This makes it "less suitable" for having some sort of programming socket, not likely as a standard product. Sane people making their own stuff would use a more DIY-friendly form factor, but that's not much comfort when the project is to replace CB2S. I suppose that you might find some sort of fiddly 3D printed jig.

**Reply by dori4n** [reply]
> This makes it "less suitable" for having some sort of programming socket, not likely as a standard product. 

Well, the point here is to program and test the module, and be able to return it under warranty, if it doesn't work, without the seller or manufacturer being able to complain with "you modified the device, damaged by user error" or something along those lines. Hence why I want a press-fit socket. I am okay with constructing my own using pogo pins, if that is what it takes, but if something already exists, I'd prefer avoiding to do that work needlessly a second time.

The pins on the back of that module as pictured on the [blakadder website](https://templates.blakadder.com/WT32C3-01N.html#flashing "Wireless Tag WT32C3-01N Module (WT32C3-01N) # Flashing") have a very interesting arrangement, but even in that picture you can see small scratches that are indicative of the pictured module having been in a socket of some sort:
<img src="https://templates.blakadder.com/assets/device_images/WT32C3-01N_pinout.webp" width="450">

I'll try building one with pogo pins, then, I guess...

**Reply by sfromis** [reply]
Still, for flashing you need to connect "IO9" too, it is not possible only on the bottom edge. That's what makes it "less straightforward" to have an "easy adapter".

The module is designed for commercial use, where manufacturers are likely to use a rather expensive automated assembly line setup connecting to the test pads (including IO9). Maybe have chips pre-programmed before even soldering to the module. Being less popular for DIY (for good reasons) makes the market for a DIY-friendly jig tiny.

**Reply by luckylinux** [reply]
What about replacing an ESP8265 Device (such as on the Sonoff 4ch pro r3) ?

Is the ESP32-C3 pin-for-pin Compatible ? Would there be a Module that fits as it is or there is a need for an Adapter Interface since the GPIOs to pull low/high during flashing seem to have somewhat been reversed ?

**Reply by sfromis** [reply]
No, the small chip (black square about 5x5 mm) is not pin compatible, and AFAICT it is soldered directly to the PCB of the Sonoff 4CH Pro R3, meaning no chance of finding a fitting module to swap. Of course, if you are good at micro-soldering, you could do the fiddly job of making an adapter.

Or just get another 4-relay board having ESP32, or maybe having ESP8266 in a ESP-12 module where you *can* find compatible replacements with ESP32-C3 (or other ESP32). Many ESP8266-based devices have a similar replacement path via compatible ESP32-based modules.
